src/mscorlib/src/System/Text/StringBuilder.cs

```diff
@@ -999,6 +999,130 @@ void ISerializable.GetObjectData(SerializationInfo info, StreamingContext contex
             return this;
         }
 
+        public StringBuilder Append(StringBuilder value)
+        {
+            Contract.Ensures(Contract.Result<StringBuilder>() != null);
+
+            if (value != null && value.Length > 0)
+            {
+                if (value != this)
```

Comment:
Here is the scenario I have in mind:

Thread 1: Create `StringBuilder sb1`
Thread 1: Create `StringBuilder sb2`
... // Append some data to both sb1 and sb2
Thread 1: Call `sb1.Append(sb2)`
Thread 2: Call `sb2.Replace("a", "B")`

At this point I think the contents of `sb1` are undetermined since while we are traversing the internal `sb2` object on thread 1, we are mutating it on thread 2.

As far as I can tell, the other methods on `StringBuilder` operate on immutable objects. This is the first API that allows a mutable object to be passed in to `StringBuilder`.

My point is that we need to be careful about how we handle this case in the implementation.